exporter/exporterhelper/internal/sized_elements_channel.go

```diff
@@ -0,0 +1,101 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package internal // import "go.opentelemetry.io/collector/exporter/exporterhelper/internal"
+
+import "sync/atomic"
+
+// sizedElementsChannel is a channel wrapper for sized elements with a capacity set to a total size of all the elements.
+// The channel will accept elements until the total size of the elements reaches the capacity.
+type sizedElementsChannel[T any] struct {
+	used *atomic.Uint64
+	cap  uint64
+	ch   chan T
+
+	// The following fields are used by the persistent queue to initialize the queue with the elements recovered from
+	// the disk. Synchronization is not required because the persistent queue doesn't access it concurrently.
+	// dequeue() will decreases the initElementsCount and initTotalSize before the used field.
+	initTotalSize     uint64
+	initElementsCount uint64
+}
+
+// newSizedElementsChannel creates a sized elements channel. Each element is assigned a size by the provided sizer.
+// chanCapacity is the capacity of the underlying channel which usually should be equal to the capacity of the queue to
+// avoid blocking the producer.
+func newSizedElementsChannel[T any](capacity int) *sizedElementsChannel[T] {
+	return &sizedElementsChannel[T]{
+		used: &atomic.Uint64{},
+		cap:  uint64(capacity),
+		ch:   make(chan T, capacity),
+	}
+}
+
+// enqueue puts the element into the queue with the given sized if there is enough capacity.
+// Returns an error if the queue is full. The callback is called before the element is committed to the queue.
+// If the callback returns an error, the element is not put into the queue and the error is returned.
+func (vcq *sizedElementsChannel[T]) enqueue(el T, size uint64, callback func() error) error {
+	if vcq.initTotalSize+vcq.used.Add(size) > vcq.cap {
+		vcq.used.Add(^(size - 1))
+		return ErrQueueIsFull
+	}
+	if callback != nil {
+		if err := callback(); err != nil {
+			vcq.used.Add(^(size - 1))
+			return err
+		}
+	}
+	vcq.ch <- el
+	return nil
+}
+
+// initBulkEnqueue puts the elements into the queue with the given size. It's used by the persistent queue to
+// initialize the queue with the elements recovered from the disk.
+func (vcq *sizedElementsChannel[T]) initBulkEnqueue(els []T, totalSize uint64) {
+	vcq.initElementsCount = uint64(len(els))
```

Comment:
Why do you need this?

Reply:
It is the same reason as in the previous comment. We need to know when the restored queue is drained